File: app/src/main/java/com/mobile/fragments/AccountDetailsFragment.java
package com.mobile.fragments;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.mobile.UserPreferences;
import com.mobile.responses.PeakPassInfo;
import com.mobile.seats.MPBottomSheetFragment;
import com.mobile.seats.SheetData;
import com.moviepass.R;

import java.text.SimpleDateFormat;

/*
 * Created by anubis on 9/2/17.
 */

public class AccountDetailsFragment extends MPFragment {

    Context context;
    View rootView, progress, accountInformation, changePassword;
    RelativeLayout shippingClick, billingClick;
    View peakPassContainer, newPeakPasses;
    TextView nextPeakPass, peakPasses;
    Activity myActivity;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_account_details, container, false);


        accountInformation = rootView.findViewById(R.id.UP);
        shippingClick = rootView.findViewById(R.id.MIDDLE);
        billingClick = rootView.findViewById(R.id.END);
        changePassword = rootView.findViewById(R.id.changePassword);
        nextPeakPass = rootView.findViewById(R.id.nextPeakPass);
        peakPassContainer = rootView.findViewById(R.id.peakContainer);
        newPeakPasses = rootView.findViewById(R.id.newPeakPassHeader);
        peakPasses = rootView.findViewById(R.id.peakPasses);
        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();
        setPeakPass();
    }

    private void setPeakPass() {
        boolean hasNewPeakPass = UserPreferences.INSTANCE.getHasNewPeakPass();
        PeakPassInfo pinfo = UserPreferences.INSTANCE.getRestrictions().getPeakPassInfo();
        peakPassContainer.setVisibility(pinfo.getEnabled() ? View.VISIBLE : View.GONE);
        if (!pinfo.getEnabled()) {
            return;
        }
        peakPasses.setText(getResources().getQuantityString(R.plurals.passes, pinfo.getPeakPasses().size(), pinfo.getPeakPasses().size()));
        if (TextUtils.isEmpty(pinfo.getNextRefillDate())) {
            nextPeakPass.setVisibility(View.GONE);
        } else {
            nextPeakPass.setVisibility(pinfo.getPeakPasses().size() == 0 ? View.VISIBLE : View.GONE);
            nextPeakPass.setText(getResources().getString(R.string.next_pass_applied, pinfo.getNextRefillDate()));
        }
        peakPassContainer.setOnClickListener(v -> {
            MPBottomSheetFragment.Companion.newInstance(
                    new SheetData(
                            getResources().getString(R.string.peak_pass),
                            getResources().getString(R.string.peak_pass_apply_bottom),
                            null,
                            pinfo.getCurrentPeakPass() == null || pinfo.getCurrentPeakPass().getExpires() == null
                                    ? null
                                    : getResources().getString(R.string.peak_pass_expires, pinfo.getCurrentPeakPass().expiresAsString()),
                            Gravity.CENTER
                    )
            ).show(getChildFragmentManager(), "");
        });
        newPeakPasses.setVisibility(hasNewPeakPass ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        shippingClick.setOnClickListener(v -> showFragment(new ProfileAccountShippingInformation()));

        accountInformation.setOnClickListener(v -> showFragment(new ProfileAccountInformation()));

        billingClick.setOnClickListener(v -> showFragment(new ProfileAccountPlanAndBilling()));

        changePassword.setOnClickListener(v -> showFragment(new ProfileAccountChangePassword()));
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        this.context = null;
    }

}

File: app/src/main/java/com/mobile/fragments/AccountDetailsFragment.kt
package com.mobile.fragments

import android.content.Context
import android.os.Bundle
import android.support.v7.util.DiffUtil
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.mobile.UserPreferences
import com.mobile.adapters.BasicDiffCallback
import com.mobile.model.UserInfo
import com.mobile.network.Api
import com.mobile.profile.*
import com.mobile.responses.UserInfoResponse
import com.moviepass.R
import dagger.android.support.AndroidSupportInjection
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.fragment_account_details.*
import javax.inject.Inject

/*
 * Created by anubis on 9/2/17.
 */

class AccountDetailsFragment : MPFragment() {

    var planResponse: UserInfoResponse? = null

    var planSub: Disposable? = null

    @Inject
    lateinit var api: Api

    val clickListener = object : ProfileClickListener {
        override fun onClick(pres: ProfilePresentation) {
            when (pres.type) {
                Profile.ACCOUNT_INFORMATION -> showFragment(ProfileAccountInformation())
                Profile.SHIPPING_ADDRESS -> showFragment(ProfileAccountShippingInformation())
                Profile.CHANGE_PASSWORD -> showFragment(ProfileAccountChangePassword())
                Profile.PLAN_AND_BILLING -> showFragment(ProfileAccountPlanAndBilling())
                else -> {
                }
            }
        }

    }

    val adapter: ProfileAdapter = ProfileAdapter(this, clickListener = clickListener)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_account_details, container, false)
    }

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onResume() {
        super.onResume()
        adapter.data = data
        fetchUserInfo()
    }

    private fun fetchUserInfo() {
        if (planResponse != null) {
            return
        }
        planSub?.dispose()
        planSub = api.getUserDataRx(UserPreferences.userId).subscribe { t1, t2 ->
            planResponse = t1
            adapter.data = data
        }
    }

    val data: ProfileData
        get() {
            val old = adapter.data?.data ?: emptyList()
            val newData = mutableListOf(
                    ProfilePresentation(
                            type = Profile.ACCOUNT_INFORMATION,
                            title = getString(R.string.account_information)
                    ),
                    ProfilePresentation(
                            type = Profile.CHANGE_PASSWORD,
                            title = getString(R.string.change_password)
                    ),
                    ProfilePresentation(
                            type = Profile.SHIPPING_ADDRESS,
                            title = getString(R.string.shipping_address)
                    ),
                    ProfilePresentation(
                            type = Profile.PLAN_AND_BILLING,
                            title = getString(R.string.plan_and_billing_info)
                    )
            )
            when (UserPreferences.restrictions.cappedPlan) {
                null -> {
                }
                else -> newData.add(ProfilePresentation(type = Profile.CAPPED_PLAN, data = planResponse))
            }
            when (UserPreferences.restrictions.peakPassInfo.enabled) {
                true -> {
                    if(newData.any { it.type==Profile.CAPPED_PLAN }) {
                        newData.add(ProfilePresentation(type=Profile.DIVIDER))
                    }
                    newData.add(ProfilePresentation(
                            type = Profile.PEAK_PASS
                    )
                    )
                }
            }
            return ProfileData(newData, DiffUtil.calculateDiff(BasicDiffCallback(old, newData)))
        }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        backButton.setOnClickListener { activity?.onBackPressed() }
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.itemAnimator = null
    }

}



